### ISSUE TITLE
Underscores in `networking.hostName` leads to `nixos-rebuild switch` breaking X11 session

### ISSUE THREAD

**Issue by Enzime** [NixOS/nixpkgs#138962]
### Describe the bug

Nearly every time I ran `nixos-rebuild switch`, my X11 session would stop working until I restarted. Some applications that were still open would still run, whereas others would stop working and I would be unable to launch any new applications.

When I try to launch any new applications, some variant of the following message (sometimes it would mention `:0` specifically):

```plaintext
Invalid MIT-MAGIC-COOKIE-1 keyUnable to init server: Could not connect: Connection refused
option parsing failed: Cannot open display: 
```

Which made me think of the `.Xauthority` file, but I didn't know how to read it and forgot about it. Months later, I finally decided to attempt to debug this issue and I stumbled upon the following command:

```
$ xauth list
bad_hostname/unix:0  MIT-MAGIC-COOKIE-1
badhostname/unix:0  MIT-MAGIC-COOKIE-1
```

This made me realize that the `nixos-rebuild switch` was causing my hostname to change and thus breaking my X11.

```
$ hostnamectl
   Static hostname: badhostname
Transient hostname: bad_hostname
         Icon name: computer-desktop
```

Compared to after a reboot:

```
   Static hostname: badhostname
         Icon name: computer-desktop
```

I've applied the simple fix of changing my hostname to `bad-hostname` :slightly_smiling_face:

### Steps To Reproduce
1. Set `networking.hostName = "bad_hostname";` in `/etc/nixos/configuration.nix`
2. `sudo nixos-rebuild switch`
3. Observe `Transient hostname` when running `hostnamectl`
4. Attempt to run any X11 programs
6. Observe no `Transient hostname` after restarting

### Notify maintainers
@0x4A6F @primeos @grahamc @vcunat

**Comment by vcunat** [comment]
Using underscore is generally a risk, as (historically) it's **not** allowed in hostnames.  https://en.wikipedia.org/wiki/Hostname#Syntax

**Comment by Enzime** [comment]
Personally wasn't aware of that, I've been using underscores in my kernel hostnames for many years now without incident until this as far as I'm aware.

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/blob/bc9b956714ed6eac5f8888322aac5bc41389defa/nixos/modules/tasks/network-interfaces.nix#L390

```
    networking.hostName = mkOption {
      default = "nixos";
      # Only allow hostnames without the domain name part (i.e. no FQDNs, see
      # e.g. "man 5 hostname") and require valid DNS labels (recommended
      # syntax). Note: We also allow underscores for compatibility/legacy
      # reasons (as undocumented feature):
      type = types.strMatching
        "^$|^[[:alnum:]]([[:alnum:]_-]{0,61}[[:alnum:]])?$";
```

it was added in 993baa587c4b82e791686f6ce711bcd4ee8ef933
https://github.com/NixOS/nixpkgs/pull/76542
@primeos

**Comment by vcunat** [comment]
Yes, the problem is that there's no clear consensus on the restrictions, so different parties have differently strict checks.  (and e.g. the DNS protocol has no restrictions at all, not even on zero bytes)

**Comment by vcunat** [comment]
... so if the underscores cause practical problems, let's be stricter and drop it from the regexp above?

**Comment by AleXoundOS** [comment]
Maybe print a warning when building NixOS configuration with underscore in `networking.hostName`?

**Comment by Artturin** [comment]
it'll be easier to just disallow them completely

**Comment by AleXoundOS** [comment]
I am still in favor of a warning.

**Comment by vcunat** [comment]
Warnings like this are annoying, unless you give a suitable option to silence them.  Otherwise it will bother you on every `nixos-rebuild` even though you've long accepted "the risk".

**Comment by AleXoundOS** [comment]
The use of underscores is not allowed by RFC, so doubtful to be something to be accepted in a long term.
<details>
  <summary>RFC952, ASSUMPTIONS</summary>
  
> 1. A "name" (Net, Host, Gateway, or Domain name) is a text string up
   to 24 characters drawn from the alphabet (A-Z), digits (0-9), minus
   sign (-), and period (.).  Note that periods are only allowed when
   they serve to delimit components of "domain style names". (See
   RFC-921, "Domain Name System Implementation Schedule", for
   background).  No blank or space characters are permitted as part of a
   name. No distinction is made between upper and lower case.  The first
   character must be an alpha character.  The last character must not be
   a minus sign or period.  A host which serves as a GATEWAY should have
   "-GATEWAY" or "-GW" as part of its name.  Hosts which do not serve as
   Internet gateways should not use "-GATEWAY" and "-GW" as part of
   their names. A host which is a TAC should have "-TAC" as the last
   part of its host name, if it is a DoD host.  Single character names
   or nicknames are not allowed.
  
</details>

**Comment by vcunat** [comment]
Well, the RFC is very old and reality is that underscores are quite commonly in use.  It's also made worse by DHCP; you might e.g. get the undescored name dynamically, not just from config.

**Comment by AleXoundOS** [comment]
Ok. The question is then how to avoid X11 session breaking or notify the user of the consequences?
